### PR TITLE
Revert #1753 and add additional tests to expose previously untested behavior

### LINKED ISSUE #1932: build.ninja generator not re-running when it should
#1753 broke CMake's test suite.  I must have neglected to run it before my approval in that PR.

I tracked it down to this example:

```bash
rm -f .ninja_deps .ninja_log
cat >build.ninja <<EOF
rule generator
  command = touch build.ninja
  generator = 1
build build.ninja: generator | implicit_generator_input
EOF
sleep 1
touch implicit_generator_input
ninja
ls -tr build.ninja implicit_generator_input
```

Prior to the change, the output is:

```
[1/1] touch build.ninja
ninja: no work to do.
implicit_generator_input
build.ninja
```

After the change, the output is:

```
ninja: no work to do.
build.ninja
implicit_generator_input
```

### PR DESCRIPTION
**Pull request by jdrouhard**
This reverts #1753 with the intention of re-landing the change with the fixes from #1933.

Additional tests are included that recreate some CMake test suite tests (see #1932). These regression tests are useful since they expose some edge-casey behavior.

Fixes #1932.

### CODE REVIEW

**Comment by jdrouhard** [comment]
cc: @jhasse @bradking

**Review by bradking — CHANGES_REQUESTED** [review]
The revert looks correct.  The test cases are representative of what we discussed in #1932 and #1933.

I've made some minor comments that need addressing.

Also, please update the revert's commit message to mention that it fixes issue #1932.

File: src/build_test.cc [hunk 1]
```diff
@@ -1589,6 +1646,32 @@ TEST_F(BuildWithLogTest, RestatMissingInput) {
   ASSERT_EQ(restat_mtime, log_entry->mtime);
 }
 
+TEST_F(BuildWithLogTest, GeneratedPlainDepfileMtime) {
+  ASSERT_NO_FATAL_FAILURE(AssertParse(&state_,
+"rule generate-depfile\n"
+"  command = echo \"$out: inimp\" > $depfile ; touch $out\n"
+"build out: generate-depfile\n"
+"  depfile = out.d\n"));
```
[the remark this answers is not in the document]
Comment (jdrouhard) [reply src/build_test.cc:1654]: I thought I'd updated all these to use the `test_dependency` binding. Must have missed this one in the cherry-pick.

[the remark this answers is not in the document]
Comment (jdrouhard) [reply src/build_test.cc:1655]: Done.

File: src/build_test.cc [hunk 2]
```diff
@@ -2574,6 +2515,90 @@ TEST_F(BuildWithDepsLogTest, DepFileOKDepsLog) {
   }
 }
 
+TEST_F(BuildWithDepsLogTest, DiscoveredDepDuringBuildChanged) {
+  string err;
+  const char* manifest =
+    "rule touch-out-implicit-dep\n"
+    "  command = touch $out ; sleep 1 ; touch $test_dependency\n"
+    "rule generate-depfile\n"
+    "  command = touch $out; echo \"$out: $test_dependency\" > $depfile\n"
```
[the remark this answers is not in the document]
Comment (jdrouhard) [reply src/build_test.cc]: done

**Review by bradking — APPROVED** [review]
